src/topology/config/validation.rs

```diff
@@ -17,278 +25,247 @@ enum Node {
-            type = "sampler"
-            inputs = ["one", "two"]
-            rate = 10
-            pass_list = []
-
-            [sinks.out]
-            type = "tcp"
-            inputs = ["three"]
-            address = "127.0.0.1:9999"
-          "#
-            .as_bytes(),
-        )
-        .unwrap();
-
-        assert_eq!(false, contains_cycle(&acyclic));
+        let mut graph = Graph::default();
+        graph.add_source("in", DataType::Log);
+        graph.add_transform("one", DataType::Log, DataType::Log, vec!["in"]);
+        graph.add_transform("two", DataType::Log, DataType::Log, vec!["in"]);
+        graph.add_transform("three", DataType::Log, DataType::Log, vec!["one", "two"]);
+        graph.add_sink("out", DataType::Log, vec!["three"]);
+
+        assert_eq!(false, graph.contains_cycle());
     }
 
     #[test]
     fn detects_type_mismatches() {
-        // Define a "minimal" Metric-typed sink so our example config can trigger a type error. As
-        // soon as we've actually implemented a Metric-typed sink or transform, we can get rid of
-        // this and just use one of those.
-        // TODO: remove all this once we have an actual non-Log component
-        use crate::{
-            buffers::Acker,
-            sinks::{
-                blackhole::{BlackholeConfig, BlackholeSink},
-                Healthcheck, RouterSink,
-            },
-            topology::config::DataType,
-        };
-        use serde::{Deserialize, Serialize};
-
-        #[derive(Deserialize, Serialize, Debug)]
-        pub struct MetricsSinkConfig;
-
-        #[typetag::serde(name = "metrics_sink")]
-        impl crate::topology::config::SinkConfig for MetricsSinkConfig {
-            fn build(&self, acker: Acker) -> Result<(RouterSink, Healthcheck), String> {
-                Ok((
-                    Box::new(BlackholeSink::new(
-                        BlackholeConfig { print_amount: 1 },
-                        acker,
-                    )),
-                    Box::new(futures::future::ok(())),
-                ))
-            }
-
-            fn input_type(&self) -> DataType {
-                DataType::Metric
-            }
-        }
-        // End of stuff to delete
-
-        let badly_typed = Config::load(
-            r#"
-            [sources.in]
-            type = "tcp"
-            address = "127.0.0.1:1235"
-
-            [sinks.out]
-            type = "metrics_sink"
-            inputs = ["in"]
-          "#
-            .as_bytes(),
-        )
-        .unwrap();
+        let mut graph = Graph::default();
+        graph.add_source("in", DataType::Log);
+        graph.add_sink("out", DataType::Metric, vec!["in"]);
 
         assert_eq!(
             Err(vec![
                 "Data type mismatch between in (Log) and out (Metric)".into()
             ]),
-            typecheck(&badly_typed)
+            graph.typecheck()
+        );
+    }
+
+    #[test]
+    fn allows_log_or_metric_into_any() {
```

Comment:
do we want to add a test around checking any for transforms as well?

Reply:
Yep, will do.